toSentry method for more meaningful data

Iazel opened the pull request:
As a Sentry user, I would like to have the possibility to add a bit more of data to an object, so that is much easier to debug one problem.  
One frequent use case is for Entities (aka Models), where knowing the ID or some other identifying mix of attributes would make much easier to reproduce and debug the problem.

dcramer commented:
I really like this idea. In Python we do ``__sentry__``, and while we dont really document or promote it well, it can be unbelievably useful. It's also possible we could do something more generic like ``->toDebugContext()``, so it could be used by other platforms. Thoughts?

dcramer commented on lib/Raven/ReprSerializer.php:
```diff
@@ -29,7 +29,11 @@ protected function serializeValue($value)
-            return 'Object '.get_class($value);
+            $repr = 'Object ' . get_class($value);
+            if (method_exists($value, 'toSentry')) {
+                $repr .= ' ' . $value->toSentry();
```
I'm thinking that if it provides toSentry maybe we just use that representation (and pass it back through the serializer). That means you could return an array of values or something similar.

stayallive replied:
Choosing the return type of `toSentry` would be great instead of it always being casted to a string, would also prevent `array to string conversion` notices that are bound to happen with this approach.

Iazel replied:
I retained the class name by default because I guess is a valuable information I think it will go in virtually any `toSentry` method; what do you think guys? Perhaps we can remove the `Object` though.

About the fact of not locking the return type to be a string, I think that is a good suggestion, will fix it on next commit

stayallive replied:
Yeah it would be great to have some sort of type indication for variables, but looking at the [docs](https://docs.sentry.io/clientdev/interfaces/stacktrace/) that currently is not possible.

Would maybe then leave it to the user to be specific about the type of the variable in their return from `toSentry`.

How is this done in Python @dcramer? Maybe the type indication for variables could be a nice addition to the Sentry server/protocol.

stayallive commented:
I personally really like `toSentry` since it so expressive (and we talk about "fixing sentries" and looking at that "sentry" instead of "debugger"), but `toDebugContext` (or both!?) would also be fine IMHO.

Iazel commented:
Thanks for the suggestions guys :+1:

 * Added support for `toDebugContext`, but I've given priority to `toSentry` in case the user want to specialize the implementation (eg: privacy reason).
 * Now the return value is also serialized, so that can be anything and not forced to a string
 * I still think that for now the `Object <class>` should be leaved in place, just to have at least this kind of information in case something goes wrong in the `toSentry` implementation :)

stayallive commented on lib/Raven/ReprSerializer.php:
```diff
@@ -29,7 +29,13 @@ protected function serializeValue($value)
+            if (method_exists($value, 'toSentry')) {
+                $repr .= ' ' . $this->serializeValue($value->toSentry());
+            } elseif (method_exists($value, 'toDebugContext')) {
+                $repr .= ' ' . $this->serializeValue($value->toSentry());
```
I think you went a bit to quick here, should be `$value->toDebugContext()` here right?

Iazel replied:
Ops, my bad xD

Iazel replied:
Fixed

stayallive commented:
Great! But now since you run it through `serializeValue` again when I return an array from `toSentry` I get `Object Namespace\MyObject Array of length 4` right? So that is not going to better (not with array's at least).

Although looking at it this was always the case for arrays... but I do get an array as array in Sentry currently, not sure why though looking at the `RepreSerializer`.

![screen shot 2017-04-13 at 13 57 18](https://cloud.githubusercontent.com/assets/1090754/25003731/27624388-2051-11e7-8f40-d84497d5a185.png)

Iazel commented:
Uhm, good point. What about type checking the value and in case is not string, either `var_export` or `json_encode` it?

dcramer commented:
@Iazel sentry can actually handle rich data types here, so as long as we return something that can eventually be json serialized we're safe (e.g. an array). If a user is providing toSentry() i think we can safely assume its a good data type.

nokitakaze commented:
You [can not](https://3v4l.org/Z5K7H) set method in stdClass, so you must declare the method in declaration. So this pull request should be done via interfaces.

```php
namespace Raven;
interface SentrySerializable {
    /*
     * @return mixed
     */
    function toSentry();
}
```

&

```php
class MyClass implements SentrySerializable {
    ...
}
```

So instead of method_exists you should check if($object instanceof \\Raven\\SentrySerializable)

This is what interfaces exist for.

nokitakaze commented:
And of course you should add tests for your code

stayallive commented:
I'm not sure what you mean @nokitakaze you cannot define any custom methods on `stdClass` and that is fine, this is for userland classes that need to be serialized how the user want too. 

Like this: https://3v4l.org/Xmmnj

A interface could definitly be used but this is and can be compatible with other error trackers without also needing the Sentry SDK.

nokitakaze commented:
> A interface could definitly be used but this is and can be compatible with other error trackers without also needing the Sentry SDK

I don't see many platforms which would use method called "toSentry".

stayallive commented:
Yes correct but for `toDebugContext` which can also be used it does make sense.

Sorry for the confusion, `toSentry` is very unlikely to be used in other SDKs.

So if we want to support both a Interface is less ideal probably...

nokitakaze commented:
If this pull request will be merged I will add new option in PR #447 which will allow using toSentry method instead of a serialization by public properties.

ste93cry commented:
Just my two cents: serializing something is always something that end-users wants to do in different ways (e.g. serialize only certain fields, use getters instead of direcly access properties, etc), so why we don't simply implement a `SerializerInterface` and then everyone is free to implement their own serializer? This way users are free to choose whatever serializer library/implementation they want to use, and most importantly we separate the real object from any code logic that handles its serialization

Iazel commented:
Hello guys, sorry but I had no time to continue this PR this week >_< That's said, thanks again for the valuable suggestions. I don't know why but I thought interfaces were introduced on php 5.3 xD

So, what changed:
* Replaced first method_exists to use a more structured interface (`Raven_Serializable`)
* Removed the 'Object <class>' prefix so now users will decide the exact value to return
* Add tests for `toSerialize` and `toDebugContext`

**Consideration**: do we want to recursively serialize if the return value is an array?

nokitakaze commented:
> interfaces were introduced on php 5.3

PHP 5.0+ gives no errors
https://3v4l.org/f0lPc

Iazel commented:
@nokitakaze Yep, after checking again I discovered that they have been introduced on php 5.0 and that's why I've used it in last commit. The important bit was:

> I don't know why but I thought ...

ie: before I thought it was like that, but than I realized it was a wrong assumption xD

Iazel commented:
@dcramer could you please check it again? Thanks :)

nokitakaze commented:
> then everyone is free to implement their own serializer

https://github.com/getsentry/sentry-php/blob/master/lib/Raven/Client.php#L197
@ste93cry, that\'s called dependency injection. And you can do a pull request which will add methods to override standard ones

nokitakaze commented on lib/Raven/Serializable.php:
```diff
@@ -0,0 +1,39 @@
+     * Can be any primitive value.
+     * In case of array, it must contains only primitive values.
+     *
+     * @return string|numeric|array
```
https://scrutinizer-ci.com/g/getsentry/sentry-php/inspections/cf2e7d62-67af-4828-9eec-bef7e2dcafaa/issues/files/lib/Raven/ReprSerializer.php?status=new&orderField=path&order=asc&honorSelectedPaths=0

As I can see toSentry may return array. But serializeValue must return "null|boolean|double|integer|string". No arrays allowed. All arrays must be serialized as well.

nokitakaze commented on test/Raven/Tests/ReprSerializerTest.php:
```diff
@@ -109,4 +109,45 @@ public function testSerializeRoundedFloat()
+        $mock->method('toDebugContext')->willReturn($expected);
+
+        $result = $serializer->serialize($mock);
+        $this->assertEquals($expected, $result);
```
Please add here a check for internal type. Only "null|boolean|double|integer|string" allowed

Iazel replied:
I guess the problem here is the fact that serializeValue is used for serializing array's keys. However the value can be a simple array and I think would be useful to allow it for more meaningfull representation.

Wouldn't be better to allow array in serializeValue and use serializeString for array's key?

nokitakaze replied:
I think only @dcramer can answer this question. This is ideology question

dcramer replied:
i dont entirely remember the logic behind how i wrote serializers, but here were the goals:

1. certain things should serialize representations of values (e.g. they "look" like the real value). This is stack args and extra context
2. Basic serialization should coerce to simple types. That should exist on our core JSON payloads, tags, etc.

ste93cry commented:
>And you can do a pull request which will add methods to override standard ones

I meant that we should provide a sort of `SerializerInterface` interface and a basic implementation (that actually can even check if the object being passed implements some methods like `toDebugContext`, but this is up to us) and if a user wants to send a custom serialization of an object then he can just (yes, using dependency injection) set the serializer to something else and do whatever he wants.

nokitakaze commented:
> SerializerInterface

I'm going to do it the other day.

HazAT commented:
Heads up, I changed the base to `1.x` since this is the old `master`.
Current `master` branch is the `2.0` version of the SDK.
Version `1.x` will receive minor and patch releases from this branch from now on.

ste93cry commented:
Closing this as it has been superseded by #809 for version `2.x`. We have no plans to implement new features in `1.x`

nokitakaze commented:
> I don't see many platforms which would use method called "toSentry".

It's funny how I changed my mind literally within a year and became strongly against **duck typing**. Now I think that an interface is primarily a code contract, not a set of signatures.
